shared.c

```diff
@@ -975,3 +989,356 @@ int appimage_unregister_in_system(char *path, gboolean verbose)
+        struct archive *a = handler->cache;
+        archive_read_close(a);
+        archive_read_free(a);
+
+        handler->cache = NULL;
+        handler->is_open = false;
+    }
+}
+
+void appimage_type1_traverse(appimage_handler * handler, traverse_cb command, void *command_data) {
+    appimage_type1_open(handler);
+
+    if (!command) {
+        fprintf(stderr, "No traverse command set.\n");
+        return;
+    }
+
+    if (handler->is_open) {
+        struct archive *a = handler->cache;
+        struct archive_entry *entry;
+        int r;
+
+        for (;;) {
+            r = archive_read_next_header(a, &entry);
+            if (r == ARCHIVE_EOF) {
+                break;
+            }
+            if (r != ARCHIVE_OK) {
+                fprintf(stderr, "%s\n", archive_error_string(a));
+                break;
+            }
+
+            /* Skip all but regular files; FIXME: Also handle symlinks correctly */
+            if(archive_entry_filetype(entry) != AE_IFREG) {
+                continue;
+            }
+
+            command(handler, entry, command_data);
+        }
+    }
+
+    appimage_type1_close(handler);
+}
+
+char* appimage_type1_get_file_name (appimage_handler * handler, void *data) {
+    struct archive_entry *entry = (struct archive_entry *) data;
+    char *filename = replace_str(archive_entry_pathname(entry), "./", "");
+    return filename;
+}
+
+void appimage_type1_extract_file (appimage_handler * handler, void *data, char *target) {
+    struct archive *a = handler->cache;
+    mk_base_dir(target);
+
+    mode_t mode = S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH;
+    int f = open(target, O_WRONLY | O_CREAT | O_TRUNC, mode);
+
+    if (f == -1){
+        fprintf(stderr, "open error: %s\n", target);
+        return;
+    }
+
+    archive_read_data_into_fd(a, f);
+    close(f);
+}
+
+appimage_handler appimage_type_1_create_handler() {
+    appimage_handler h;
+    h.traverse = appimage_type1_traverse;
+    h.get_file_name = appimage_type1_get_file_name;
+    h.extract_file = appimage_type1_extract_file;
+
+    return h;
+}
+
+/*
+ * AppImage Type 2 Functions
+ */
+
+void appimage_type2_open(appimage_handler * handler) {
+    if (is_handler_valid(handler) && !handler->is_open) {
+        fprintf(stderr, "Opening %s as Type 2 AppImage\n", handler->path);
+        long unsigned int fs_offset; // The offset at which a squashfs image is expected
+        fs_offset = get_elf_size(handler->path);
+
+        sqfs *fs = malloc(sizeof(sqfs));
+        sqfs_err err = sqfs_open_image(fs, handler->path, fs_offset);
+        if (err != SQFS_OK){
+            fprintf(stderr, "sqfs_open_image error: %s\n", handler->path);
+            sqfs_destroy(fs);
+        } else {
+            handler->is_open = true;
+            handler->cache = fs;
+        }
+    }
+}
+
+void appimage_type2_close(appimage_handler * handler) {
+    if ( is_handler_valid(handler) && handler->is_open ) {
```

Comment:
Code style: `if (is_...` or `if(is...)`. I've spotted this in many of the new functions.

Reply:
What do you exactly propose here ?

Reply:
The space inside the bracket isn't really used in the rest of the file, as far as I can see. But I don't really mind the current state either. Personally, I prefer `if (...)` over `if(...)`, but the latter is the pattern used in the majority of the code files.